### ISSUE TITLE
PBKDF from FDE proposal

### ISSUE THREAD

**Issue by woodbe** [DSC-iTC/cPP#124]
The CCDB Crypto Working group struggled with this SFR for a while and was working with the DSC iTC to figure out a solution. This is what the WG currently is proposing:

FCS_CKM_EXT.8.1 The TSF shall perform password-based key derivation functions in accordance with a specified cryptographic algorithm [HMAC-[selection: SHA-256, SHA-384, SHA-512]], with iteration count of [assignment: number of iterations] using a randomly generated salt of length [assignment: length of salt] and output cryptographic key sizes [selection: 128, 192, 256] bits that meet the following standard: [NIST SP 800-132].

There's a lot of guidance text, but we wanted to keep the SFR focused on aspects of the standard (which only talks about number of iterations and length of the salt), rather than include aspects out side of the standard (e.g., anti-hammering).

-----

This is related to [DSC0001](https://dsc-itc.github.io/TD/DSC0001.html) and would be a further change in v1.1.

**Comment by woodbe** [comment]
This should probably be handled by creating a separate anti-hammering SFR and then setting that SFR to have a statement that it could be handled by either PBKDF rounds or by some other means, and leave those to be set in the new SFR.

This would be possible for the new version while it was not in the TD.

**Comment by woodbe** [comment]
Thinking something along the lines of https://www.niap-ccevs.org/MMO/PP/MDF%203.3%20PP/#FIA_TRT_EXT.1

This may be selection-based (along with the FCS_CKM_EXT.8, or could be mandatory).

**Comment by woodbe** [comment]
So one question about this is what about other key expansion techniques? For example, in MDF there is an open option to specify another key expansion type that doesn't meet a specific standard. Android uses scrypt rounds surrounded by PBKDF2 iterations to do this, but in the DSC as this is, if I only used scrypt instead of PBKDF, I wouldn't be able to claim this requirement (its optional, so maybe that is OK, but it does seem somewhat limiting).

@jdonndelinger what do you think?

**Comment by woodbe** [comment]
Trying to think about this issue to close out crypto requirements. What are we trying to solve with all the adjustments like [DSC0001](https://dsc-itc.github.io/TD/DSC0001.html) and the [FDE change](https://www.niap-ccevs.org/Documents_and_Guidance/view_td.cfm?TD=0764)?

The way the new catalog SFR is written is very simply with no restrictions about the number of iterations (which really begs the question as to why it isn't FCS_COP.1/PBKDF like it was before, since the FCS_CKM would imply more than just an algorithm), but the app note is heavy on talking about password quality and binding to the TOE.

This proposal originally was to add an anti-hammering requirement similar to the PP_MDF, but I'm not sure that is really needed here. There is already an explicit auth counter in FIA_AFL_EXT.1, which as an extended SFR could be adjusted to include a time limit on attempts, but this doesn't seem to really be an issue for the type of system the DSC is. Brute force hammering on the device doesn't seem like the place where attacks are going to happen, it really seems like an offline version, where the stored data has somehow been exfiltrated from the device, is the most likely attack, and while PBKDF can help, it would seem that TOE binding is the better answer here.

What about some sort of SFR that looks like this (this would be used when FCS_STG_EXT.1.1 selects "software-based"):

> FCS_KDB_EXT.1 Key-to-Device Binding
FCS_KDB_EXT.1.1 The TSF shall bind all KEKs to a Root Encryption Key of the TOE.

This would specifically mandate then, for software-based key storage that the data be explicitly bound to the device to prevent offline attacks by anything other than brute force guessing of the encryption key (even if they somehow had the authorization data). 

I do not actually see anything like this specified anywhere. It is sort of hinted at, but not stated. If this is used, then it would seem that the PBKDF can just be "how many rounds are you using". 

I'm not exactly sure if this is sufficient if the device uses mutable hardware storage, since there we expect the data is not encrypted but is protected by hardware, and hence not accessible. In that case the binding doesn't get anything, but I'm not clear that this matters.

**Comment by smuellerDD** [comment]
The key problem that the change shall solve is that the anti-hammering protection shall be allowed to be delivered by another means than PBKDF. Thus, if the PBKDF round count is very low, then surely there must be another anti-hammering protection. And the ST shall explain that to show it is sufficiently strong at least en par with the original PBKDF anti-hammering requirement.

And, the anti-hammering IMHO is truly needed even to counter online attacks. The reason is that the offline attack you mention requires the offloading of the stored PBKDF-ed version of the passcode. For a DSC, I hope that should be prohibited by the pure nature of the TOE. Yet, the PBKDF operation may be used as central OS authentication mechanism (see Apple). In this case, you want a proper PBKDF anti-hammering solution to counter online attacks. Thus, the binding to a hardware-key will not solve the online-anti-hammering solution.

But perhaps that distinction online vs offline anti-hammering protection can be used here. I.e. for offline protection, your suggestion is wise. For online-protection, the mentioned SFR text should be applied.

**Comment by robertmclemons** [comment]
I agree. If password-derived keys are bound to the TOE such that only the TOE can decrypt using such a key, then the number of rounds to derive the key is not very important if hammering attempts can be detected or rendered impractical. through use of timers or counters.

If password-derived keys are not bound to the TOE, then the number of rounds needs to be very high to inconvenience an off-line attacker.

**Comment by woodbe** [comment]
But in the Apple solution they didn't use PBKDF rounds to do the protection, so for online protection this still, straight-up, doesn't resolve the issue. I don't have an issue with using PBKDF rounds to provide hammering protection, but it isn't the only way to do so. Limits on number of attempts over time, or other methods that eat time/resources (such as scrypt which is used in Android, but not at this layer, but still), are non-PBKDF methods that perform the same function. 

It still seems like we need to specify some sort of separate hammering SFR. Along the lines of:

> The TSF shall use [selection: PBKDF rounds above 1000, a maximum attempts counter of [assignment: attempts per time period], [assignment: other method]] to prevent hammering attempts to access keys protected by the TOE.

Then if PBKDF is used, you need to see some number of rounds in that statement, but if something else is used, then minimal PBKDF rounds can be specified.

**Comment by robertmclemons** [comment]
Yeah. I think CWG didn't want to mix the key derivation function with non-crypto requirements. I don't think we need to be constrained by such concerns, but we do need a requirement to invoke PBKDF, and maybe that one can talk about anti-hammer and rounds. Or I could just be overcomplicating things.

**Comment by smuellerDD** [comment]
If you want to change it, all I would like to ensure is that PBKDF2 with one round plus a implementation-specific anti-hammering is allowed.
